cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -0,0 +1,343 @@
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+    Set<Replica> candidateReplicas = _selfHealingDeadBrokersOnly ? new HashSet<>(broker.replicas())
+                                                                 : new HashSet<>(broker.leaderReplicas());
+    if (!clusterModel.deadBrokers().isEmpty() && broker.isAlive()) {
+      candidateReplicas.retainAll(broker.immigrantReplicas());
+    }
+    int numReplicas = candidateReplicas.size();
+    for (Replica replica : candidateReplicas) {
+      if (shouldExclude(replica, excludedTopics)) {
+        continue;
+      }
+      Broker b = maybeApplyBalancingAction(clusterModel,
+                                           replica,
+                                           candidateBrokers,
+                                           ActionType.INTER_BROKER_REPLICA_MOVEMENT,
+                                           optimizedGoals,
+                                           optimizationOptions);
+      // Only check if we successfully moved something.
+      if (b != null) {
+        if (--numReplicas <= balanceUpperLimit) {
+          return false;
+        }
+        // Remove and reinsert the broker so the order is correct.
+        candidateBrokers.remove(b);
+        if (b.leaderReplicas().size() < _balanceUpperLimit || _selfHealingDeadBrokersOnly) {
+          candidateBrokers.add(b);
+        }
+      }
+    }
+    return true;
+  }
+
+  private boolean rebalanceByMovingLeaderReplicasIn(Broker broker,
+                                                   ClusterModel clusterModel,
+                                                   Set<Goal> optimizedGoals,
+                                                   OptimizationOptions optimizationOptions) {
+    if (optimizationOptions.excludedBrokersForLeadership().contains(broker.id())) {
+      return true;
+    }
+
+    PriorityQueue<Broker> eligibleBrokers = new PriorityQueue<>((b1, b2) -> {
+      int result = Integer.compare(b2.leaderReplicas().size(), b1.leaderReplicas().size());
+      return result == 0 ? Integer.compare(b1.id(), b2.id()) : result;
+    });
+
+    for (Broker aliveBroker : clusterModel.aliveBrokers()) {
+      if (aliveBroker.leaderReplicas().size() > _balanceLowerLimit) {
+        eligibleBrokers.add(aliveBroker);
+      }
+    }
+    List<Broker> candidateBrokers = Collections.singletonList(broker);
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+    int numLeaderReplicas = broker.leaderReplicas().size();
+    while (!eligibleBrokers.isEmpty()) {
+      Broker sourceBroker = eligibleBrokers.poll();
+      for (Replica replica : new HashSet<>(sourceBroker.leaderReplicas())) {
+        if (shouldExclude(replica, excludedTopics) || broker.replica(replica.topicPartition()) != null) {
+          continue;
+        }
+        if (!clusterModel.deadBrokers().isEmpty() || !broker.immigrantReplicas().contains(replica)) {
```

Comment:
Would you clarify what this check means?
It looks like the current code skips moving leader replicas in if (1) there is a dead broker in the cluster or (2) the leader replica that is intended to be moved to this broker (an currently resides on another broker) is not an immigrant replica in this broker (which I think will always be `false`).

Reply:
Good catch! 
It should be `!clusterModel.deadBrokers().isEmpty() && !sourceBroker.immigrantReplicas().contains(replica)`, which limit the movement to only immigrant replica in self-healing case.